### ISSUE TITLE
Osmosis version 0.45

### ISSUE THREAD

**Issue by colinrlly** [locked-fg/osmosis-srtm-plugin#11]
I am trying to use this plugin with Osmosis version 0.45. This version has plexus-classworlds-2.5.2.jar in lib/default. In the instructions it says I may need to adjust bin/osmosis.bat. After installing and running Osmosis using --write-srtm the created .xml file includes <tag k="height" v="NaN"/> on all the nodes. Can you speak more to what needs to be adjusted in the .bat file?

**Comment by locked-fg** [comment]
It's quite some time ago. 
But back then the only adjustment was to ensure the correct version numbers. 

But if you say that the output worked... This does not sound as if you'd need to change anything?

**Comment by colinrlly** [comment]
Thanks for the quick response. What I mean is every node in my osm file looks like this:

<node id="29568783" version="5" timestamp="2014-11-09T16:14:57Z" uid="2311536" user="JordanKepler" changeset="26666712" lat="34.8478941" lon="-82.358738">
    <tag k="height" v="NaN"/>
  </node>
  <node id="29568784" version="5" timestamp="2014-11-09T16:14:57Z" uid="2311536" user="JordanKepler" changeset="26666712" lat="34.84785" lon="-82.3580849">
    <tag k="height" v="NaN"/>
  </node>

None of them include an actual height value. When I say the output worked I mean the program didn't explode, but it didn't give the desired output either.

**Comment by colinrlly** [comment]
Sorry I'm new to making issues on Github, I see my code example was never included in my original post.

**Comment by locked-fg** [comment]
Hm that's strange.
Just to be sure, did you try both directories and create / modify the .conf -file?:
- Create  bin\plugins and put the build/downloaded jar there (**if this doesn't work, try lib\default**)
- create a osmosis-plugins.conf in the config-folder containing the following line de.locked.osmosis.srtmplugin.SrtmPlugin_loader

Maybe osmosis has changed the way the load plugins. If the above doesn't work, can you maybe get an older osmosis-version and try?

**Comment by locked-fg** [comment]
How's it going? did you manage to solve the issue?

**Comment by colinrlly** [comment]
I tried putting the jar in bin/plugins and lib/default. I have osmosis-plugins.conf in osmosis-latest/config with that line in in. In both scenarios I get the same output:

```
stderr: Sep 01, 2017 7:29:22 PM org.openstreetmap.osmosis.core.Osmosis run
INFO: Osmosis Version 0.45

stderr: Sep 01, 2017 7:29:22 PM org.openstreetmap.osmosis.core.Osmosis run
INFO: Preparing pipeline.

stderr: Sep 01, 2017 7:29:22 PM org.openstreetmap.osmosis.core.Osmosis run
INFO: Launching pipeline execution.

stderr: Sep 01, 2017 7:29:22 PM org.openstreetmap.osmosis.core.Osmosis run
INFO: Pipeline executing, waiting for completion.

stderr: Sep 01, 2017 7:29:23 PM org.openstreetmap.osmosis.core.Osmosis run
INFO: Pipeline complete.

stderr: Sep 01, 2017 7:29:23 PM org.openstreetmap.osmosis.core.Osmosis run
INFO: Total execution time: 1450 milliseconds.

osmosis exited with code 0
```

My code looks like this. I'm using Node.js and spawning a child process that runs osmosis.bat.

```
var osmread = require('osm-read');
var cp = require('child_process')
var osmosis = require.resolve('./osmosis-latest/bin/osmosis.bat');
var bat = cp.spawn(
    osmosis, 
    ['--read-xml', '../../../static/burlington.xml', '--write-srtm', '--write-xml', '../../../static/burlington_with_srtp.xml'],
    {cwd: './osmosis-latest/bin'}
);

bat.stdout.on('data', function (data) {
    console.log('stdout: ' + data);
});

bat.stderr.on('data', function (data) {
    console.log('stderr: ' + data);
});

bat.on('exit', function (code) {
    console.log('osmosis exited with code ' + code);
});
```

Like I said before it has to be working somewhat because the program does modify my osm data to include `<tag k="height" v="NaN"/>` in every node.

**Comment by locked-fg** [comment]
Ah this was a good hint! Can you please first try if it just works in the console without NodeJS?
Just to test if its an issue with osmosis or with the environment provided by NodeJS.

**Comment by locked-fg** [comment]
At least I don't recall about it. Sorry that I currently cannot actively dig into the problem deeper.
SRTM files were downloaded / are locally available?

**Comment by colinrlly** [comment]
I need to have the SRTM files on my computer? I have the osm .xml files but I didn't know I need to have the SRTM .hgt files. How do I tell the Osmosis plugin to use the SRTM files on my computer?

**Comment by locked-fg** [comment]
Nono, they are (they should!) be downloaded automatically. I just wonder if there is a download issue (server url changed, anti virus blocking download, corrupt files) that caused the NANs.
Judging the code in https://github.com/locked-fg/osmosis-srtm-plugin/blob/master/src/main/java/de/locked/osmosis/srtmplugin/SrtmPlugin_task.java
file issues cause return Double.NaN quite often.

Do you see a log file created during the process? Errors should be logged there.

**Comment by thyrfa** [comment]
I'm encountering the same issue, all nodes are returning NaN regardless of if I use local files or not.  Sidenote, in Osmosis 0.46 the run fails with the message 

> SEVERE: Thread for task 1-read-xml failed
> java.lang.AbstractMethodError: de.locked.osmosis.srtmplugin.SrtmPlugin_task.close()V
>         at org.openstreetmap.osmosis.xml.v0_6.XmlReader.run(XmlReader.java:107)
>         at java.base/java.lang.Thread.run(Thread.java:844)
> 
but the output file does get generated anyways (with NaNs).  If I run it in Osmosis 0.43.1 it doesn't give an error message but still fills every node with NaN.

**Comment by locked-fg** [comment]
Strange. If you're already in building the JAR yourself: can you maybe add some logging to see where the NANs are returned? I'd need to set up everything as I haven't touch this things for quite some time to be honest

**Comment by thyrfa** [comment]
Ok: I got it to work.  Takeaways: 

1) The server downloading functionality does not work anymore.  It will return error.  Not a huge priority if you just direct users to the website to manually download files AND
2) Make sure the users have .hgt.zip, I didnt realize this and unzipped them (Whoops).
3) The prebuilt version you link to is behind the current live github version.  The live version does not write to the height tag if the value is NaN while the prebuilt version does.  (Explains my confusion).
4) Regardless of using Java 8 or Java 9 (Figured I would check both) Osmosis 0.46 will return the below error when using the plugin.  I recommend 0.43.1 for compatibilities sake, but even though Osmosis 0.46 returns the below error, **the data still gets written correctly**

> SEVERE: Thread for task 1-read-xml failed
> java.lang.AbstractMethodError: de.locked.osmosis.srtmplugin.SrtmPlugin_task.close()V
>         at org.openstreetmap.osmosis.xml.v0_6.XmlReader.run(XmlReader.java:107)
>         at java.lang.Thread.run(Thread.java:748)
> 
> Nov 13, 2017 6:39:50 PM org.openstreetmap.osmosis.core.Osmosis main
> SEVERE: Execution aborted.
> org.openstreetmap.osmosis.core.OsmosisRuntimeException: One or more tasks failed.
>         at 
> org.openstreetmap.osmosis.core.pipeline.common.Pipeline.waitForCompletion(Pipeline.java:146)
Truncated for space

In Summary: It works as long as users go manually get the files and put them in a folder, making sure all the .hgt files stay zipped.

**Comment by thyrfa** [comment]
Made a PR for the README so others will know the workaround.

**Comment by locked-fg** [comment]
Thanks a lot! At least there is a workaround!

**Comment by redonggaojianbing** [comment]
So, Thyrfa. How did you reslove the problem that all the node provide NaN height tags?

**Comment by thyrfa** [comment]
Like it clearly says above: 

> It works as long as users go manually get the files and put them in a folder, making sure all the .hgt files stay zipped.


Go through the steps in the readme and you should be fine.

**Comment by redonggaojianbing** [comment]
It goes well. Thank you so much！
